Does not work with java 16+?

Hi,

We are very happy users of your library, many thanks for this awesome piece of code.

We are however having initialization issues on machines using java 16 and above, which seems to be caused by a change in nio.Buffer. A StackTrace is below. Do you have an idea on how to circumvent this?

Many thanks for your help,

Marc


```
Exception in thread "ImportThread" java.lang.ExceptionInInitializerError
	at io.airlift.compress.zstd.ZstdFrameCompressor.writeMagic(ZstdFrameCompressor.java:57)
	at io.airlift.compress.zstd.ZstdFrameCompressor.compress(ZstdFrameCompressor.java:143)
	at io.airlift.compress.zstd.ZstdCompressor.compress(ZstdCompressor.java:45)
	...
        Suppressed: java.lang.NoClassDefFoundError: Could not initialize class io.airlift.compress.zstd.UnsafeUtil
		at io.airlift.compress.zstd.ZstdFrameCompressor.writeMagic(ZstdFrameCompressor.java:57)
		at io.airlift.compress.zstd.ZstdFrameCompressor.compress(ZstdFrameCompressor.java:143)
		at io.airlift.compress.zstd.ZstdCompressor.compress(ZstdCompressor.java:45)
		...
Caused by: io.airlift.compress.IncompatibleJvmException: Zstandard requires access to java.nio.Buffer raw address field
	at io.airlift.compress.zstd.UnsafeUtil.<clinit>(UnsafeUtil.java:53)
	...
```

--- a/src/main/java/io/airlift/compress/lz4/UnsafeUtil.java
+++ b/src/main/java/io/airlift/compress/lz4/UnsafeUtil.java
@@ -25,7 +25,7 @@ import static java.lang.String.format;
 final class UnsafeUtil
 {
     public static final Unsafe UNSAFE;
-    private static final Field ADDRESS_ACCESSOR;
+    private static final long ADDRESS_OFFSET;
 
     private UnsafeUtil() {}
 
@@ -45,22 +45,20 @@ final class UnsafeUtil
         }
 
         try {
-            Field field = Buffer.class.getDeclaredField("address");
-            field.setAccessible(true);
-            ADDRESS_ACCESSOR = field;
+            // fetch the address field for direct buffers
+            ADDRESS_OFFSET = UNSAFE.objectFieldOffset(Buffer.class.getDeclaredField("address"));
         }
-        catch (Exception e) {
+        catch (NoSuchFieldException e) {
             throw new IncompatibleJvmException("LZ4 requires access to java.nio.Buffer raw address field");
         }
     }
 
     public static long getAddress(Buffer buffer)
     {
-        try {
-            return (long) ADDRESS_ACCESSOR.get(buffer);
+        if (!buffer.isDirect()) {
+            throw new IllegalArgumentException("buffer is not direct");
         }
-        catch (IllegalAccessException e) {
-            throw new RuntimeException(e);
-        }
+
+        return UNSAFE.getLong(buffer, ADDRESS_OFFSET);
     }
 }

--- a/src/main/java/io/airlift/compress/lzo/UnsafeUtil.java
+++ b/src/main/java/io/airlift/compress/lzo/UnsafeUtil.java
@@ -25,7 +25,7 @@ import static java.lang.String.format;
 final class UnsafeUtil
 {
     public static final Unsafe UNSAFE;
-    private static final Field ADDRESS_ACCESSOR;
+    private static final long ADDRESS_OFFSET;
 
     private UnsafeUtil() {}
 
@@ -45,22 +45,20 @@ final class UnsafeUtil
         }
 
         try {
-            Field field = Buffer.class.getDeclaredField("address");
-            field.setAccessible(true);
-            ADDRESS_ACCESSOR = field;
+            // fetch the address field for direct buffers
+            ADDRESS_OFFSET = UNSAFE.objectFieldOffset(Buffer.class.getDeclaredField("address"));
         }
-        catch (Exception e) {
+        catch (NoSuchFieldException e) {
             throw new IncompatibleJvmException("LZO requires access to java.nio.Buffer raw address field");
         }
     }
 
     public static long getAddress(Buffer buffer)
     {
-        try {
-            return (long) ADDRESS_ACCESSOR.get(buffer);
+        if (!buffer.isDirect()) {
+            throw new IllegalArgumentException("buffer is not direct");
         }
-        catch (IllegalAccessException e) {
-            throw new RuntimeException(e);
-        }
+
+        return UNSAFE.getLong(buffer, ADDRESS_OFFSET);
     }
 }

--- a/src/main/java/io/airlift/compress/snappy/UnsafeUtil.java
+++ b/src/main/java/io/airlift/compress/snappy/UnsafeUtil.java
@@ -25,7 +25,7 @@ import static java.lang.String.format;
 final class UnsafeUtil
 {
     public static final Unsafe UNSAFE;
-    private static final Field ADDRESS_ACCESSOR;
+    private static final long ADDRESS_OFFSET;
 
     private UnsafeUtil() {}
 
@@ -45,22 +45,20 @@ final class UnsafeUtil
         }
 
         try {
-            Field field = Buffer.class.getDeclaredField("address");
-            field.setAccessible(true);
-            ADDRESS_ACCESSOR = field;
+            // fetch the address field for direct buffers
+            ADDRESS_OFFSET = UNSAFE.objectFieldOffset(Buffer.class.getDeclaredField("address"));
         }
-        catch (Exception e) {
+        catch (NoSuchFieldException e) {
             throw new IncompatibleJvmException("Snappy requires access to java.nio.Buffer raw address field");
         }
     }
 
     public static long getAddress(Buffer buffer)
     {
-        try {
-            return (long) ADDRESS_ACCESSOR.get(buffer);
+        if (!buffer.isDirect()) {
+            throw new IllegalArgumentException("buffer is not direct");
         }
-        catch (IllegalAccessException e) {
-            throw new RuntimeException(e);
-        }
+
+        return UNSAFE.getLong(buffer, ADDRESS_OFFSET);
     }
 }

--- a/src/main/java/io/airlift/compress/zstd/UnsafeUtil.java
+++ b/src/main/java/io/airlift/compress/zstd/UnsafeUtil.java
@@ -25,7 +25,7 @@ import static java.lang.String.format;
 final class UnsafeUtil
 {
     public static final Unsafe UNSAFE;
-    private static final Field ADDRESS_ACCESSOR;
+    private static final long ADDRESS_OFFSET;
 
     private UnsafeUtil() {}
 
@@ -45,22 +45,20 @@ final class UnsafeUtil
         }
 
         try {
-            Field field = Buffer.class.getDeclaredField("address");
-            field.setAccessible(true);
-            ADDRESS_ACCESSOR = field;
+            // fetch the address field for direct buffers
+            ADDRESS_OFFSET = UNSAFE.objectFieldOffset(Buffer.class.getDeclaredField("address"));
         }
-        catch (Exception e) {
+        catch (NoSuchFieldException e) {
             throw new IncompatibleJvmException("Zstandard requires access to java.nio.Buffer raw address field");
         }
     }
 
     public static long getAddress(Buffer buffer)
     {
-        try {
-            return (long) ADDRESS_ACCESSOR.get(buffer);
+        if (!buffer.isDirect()) {
+            throw new IllegalArgumentException("buffer is not direct");
         }
-        catch (IllegalAccessException e) {
-            throw new RuntimeException(e);
-        }
+
+        return UNSAFE.getLong(buffer, ADDRESS_OFFSET);
     }
 }

